### ISSUE TITLE
Memory leak in finalizing GeneratorDataset iterator

### ISSUE THREAD

**Issue by Mauhing** [tensorflow/tensorflow#36919]
**System information**
- OS Linux Ubuntu 18.04  
- TensorFlow installed from `pipenv install tensorflow~=2.1  
- TensorFlow version 2.1.0
- Python version: 3.6
- CUDA/cuDNN version: 10.1
- GPU model and memory: 2 x (RTX 2080 Ti 11GB) 

**Describe the current behavior**

`tf.keras.preprocessing.image.ImageDataGenerator` output warning message `Error occurred when finalizing GeneratorDataset iterator: Cancelled: Operation was canceled` when I am using `tf.keras.model.fit(..., worker = 10)` start to train to next epoch. The warning is gone when `worker = 1`. However, the warning does not seem to affect the result. 

In #35100, people tend to mean the warning is superfluous and spurious and the warning has been removed in tf-nightly. But after my usage case, I believe the warning is correct. I have infinite models to be trained one by one (generated by a genetic algorithm). When `tf.keras.model.fit(..., worker = 10)` the error message print out and the memory does not get freed in the next epoch. After training a few models.  All the memory is occupied and the process gets killed. It does not happen when `worker=1` the memory gets freed for each epoch.

I have try tf-nightly, The error message is removed but the memory problem is still there. 

**Describe the expected behavior**
It should free the memory for every epoch. 

**Code to reproduce the issue** 
I prepare my data in the following way.
`samples` is NumPy array with shape (n, 90, 90, 3). n'th image withe 90 by 90 RGB image.
`labels` is NumPy array with shape (n, 2), which is yes and no. [1, 0] is yes and [0, 1] is no.

I have to make a pipenv file, data set and small script for you to demonstrate the memory problem. 
[Download link from google drive](https://drive.google.com/file/d/1oQ1wQkY1ZXimQNBQljlMbNB4VNYjDAp7/view?usp=sharing)
(300 MB)

Unzip the link above and run
`pipenv install`
`pipenv shell`
`python main.py`
Then you see the warning

``` python
train_generator = data_aug.flow(samples, labels, batch_size = 64)
self.model.fit(train_generator, epochs=70, max_queue_size = 64, workers = 10)
```

**Other info / logs**
During these 70 epochs, you can see the memory go up for each epoch and print the error message every time between the epoch.

**Comment by ravikyram** [comment]
@Mauhing 

Could you please share colab link or simple standalone code with supporting files to reproduce the issue in our environment. It helps in localizing the issue faster. Thanks!

**Comment by Mauhing** [comment]
> @Mauhing
> 
> Could you please share colab link or simple standalone code with supporting files to reproduce the issue in our environment. It helps in localizing the issue faster. Thanks!

I have to make a pipenv file, data set and small script for you to demonstrate the memory problem. 
[https://drive.google.com/file/d/1oQ1wQkY1ZXimQNBQljlMbNB4VNYjDAp7/view?usp=sharing](url)
(300 MB)

Unzip the link above and run
`pipenv install`
`pipenv shell`
`python run`
Then you see the warning

The problem occurs when you train it on GPU. The warning shown twice when you train on two GPU at each epoch.

**Comment by gowthamkpr** [comment]
Was able to reproduce the issue on colab. Please find the gist [here](https://colab.sandbox.google.com/gist/gowthamkpr/3fddb14893a157806d399f7cb146a9f7/untitled19.ipynb). Thanks!

**Comment by Mauhing** [comment]
> Was able to reproduce the issue on colab. Please find the gist [here](https://colab.sandbox.google.com/gist/gowthamkpr/3fddb14893a157806d399f7cb146a9f7/untitled19.ipynb). Thanks!

I want to make a kindly reminder that the colab code does not seem to print out the memory leak in each epoch. But the memory does not get free in each epoch.  Thanks

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
